Expose cdrom in UI - Story #1, #2, #3, #4

glekner opened the pull request:
This PR implements a CDRom interface which includes a detail view of the CDRoms state, and a Modal to edit them.

glekner commented:
> @glekner looking good so far. Will this PR cover editing as well?

Why do you mean by editing? I have some questions about the form validations maybe you can @ me at slack or chat

glekner commented:
*Update*
* CDs type can now be changed from any type to any other type.
* Add CDRom button will be available only if the CDs count is less than 2
* The modal is capable of showing more than 2 CDs, and changing them.
* The modal has validations for the CD names and URL if needed, the form cannot be submitted without validation
* The design uses the new PF4 Modal which isn't currently implemented in the modal infrastructure, so I used what we have @matthewcarleton. @jelkosz do you think a name column is needed? and if not how should we call new CDS. 

@rawagner @jelkosz could you guys play with the branch?

matthewcarleton commented:
@glekner thanks for this! 
I'm not sure I see any value in a name when we are just pulling in from another source. Can we auto populate a name (if it's needed) from the source? So .iso, .qcow2, etc? 
I believe the need for the second CD here is strictly for windows VMs right? Can we drop it otherwise? Is there any need for a second one for any other type of VM?

rawagner commented on frontend/packages/kubevirt-plugin/src/selectors/selectors.ts:
```diff
@@ -12,6 +12,9 @@ export const getAnnotations = (vm: VMLikeEntityKind, defaultValue?: any) =>
 export const getDescription = (vm: VMLikeEntityKind) =>
   _.get(vm, 'metadata.annotations.description');
 
+export const getProvisionSource = (vm: VMLikeEntityKind) =>
```
not used anywhere

rawagner commented on frontend/packages/kubevirt-plugin/src/components/modals/cd-rom-modal/_cd-rom-modal.scss:
```diff
@@ -0,0 +1,8 @@
+.cd-action {
```
all rules should be prefixed with `kubevirt-`

rawagner commented on frontend/packages/kubevirt-plugin/src/components/modals/cd-rom-modal/cd-rom-modal.tsx:
```diff
@@ -0,0 +1,172 @@
+  ModalComponentProps,
+} from '@console/internal/components/factory';
+import { PlusCircleIcon } from '@patternfly/react-icons';
+import { Button } from 'patternfly-react';
```
can you use `Button` from PF4 ?

rawagner commented on frontend/packages/kubevirt-plugin/src/components/modals/cd-rom-modal/cd-rom-modal.tsx:
```diff
@@ -0,0 +1,172 @@
+import { initialDisk } from './constants';
+import './_cd-rom-modal.scss';
+
+export const VMCdRomModal = withHandlePromise((props: VMCdRomModalProps) => {
```
```suggestion
export const VMCdRomModal = withHandlePromise(({
  vm,
  handlePromise,
  inProgress,
  errorMessage,
  persistentVolumeClaims,
  cancel,
  close,
}: VMCdRomModalProps) => {
```

and there is no need to deconstruct later

rawagner commented on frontend/packages/kubevirt-plugin/src/components/modals/cd-rom-modal/cd-rom-modal.tsx:
```diff
@@ -0,0 +1,172 @@
+      };
+
+      const container = getContainerImageByDisk(vm, cd.name);
+      if (container) return { ...cd, type: 'container', container };
```
I prefer 
```suggestion
      if (container) {
        return { ...cd, type: 'container', container };
      }
```
to avoid mistakes in future. Applies to other `ifs` too

rawagner commented on frontend/packages/kubevirt-plugin/src/components/modals/cd-rom-modal/cd-rom-modal.tsx:
```diff
@@ -0,0 +1,172 @@
+          } else cd[key] = value;
+
+          if (key === 'url') {
+            if (!value.startsWith('http://')) cd.isURLValid = false;
```
you should reuse validation from web-ui-components `validateURL`

rawagner commented on frontend/packages/kubevirt-plugin/src/components/modals/cd-rom-modal/cd-rom-modal.tsx:
```diff
@@ -0,0 +1,172 @@
+          {cds.length > 0
+            ? cds.map((cd, i) => (
+                <CDRomRow
+                  key={`row-cd-${i}`} // eslint-disable-line react/no-array-index-key
```
`key` can be volume name as that is unique

rawagner commented on frontend/packages/kubevirt-plugin/src/components/modals/cd-rom-modal/cd-rom-modal.tsx:
```diff
@@ -0,0 +1,172 @@
+              className="btn-link--no-btn-default-values add-cdrom-btn"
+              onClick={onCDAdd}
+            >
+              <PlusCircleIcon style={{ marginRight: 'var(--pf-global--spacer--xs)' }} />
```
`style` should be extracted to `scss`

rawagner commented on frontend/packages/kubevirt-plugin/src/components/modals/cd-rom-modal/index.tsx:
```diff
@@ -0,0 +1,35 @@
+import { VirtualMachineModel } from '../../../models';
+import { VMCdRomModal } from './cd-rom-modal';
+
+const CDVModalFirehose: React.FC<CloneVMModalFirehoseProps> = (props) => {
```
why `CDV` ?

glekner replied:
suppose to be CDVMModal..
do you have a better name in mind?

rawagner replied:
I'd go with `CDRomModal`

rawagner commented on frontend/packages/kubevirt-plugin/src/selectors/vm/selectors.ts:
```diff
@@ -87,3 +89,20 @@ export const getFlavorDescription = (vm) => {
+export const getWindowsToolsURLByDisk = (vm: VMKind, name: string) =>
+  isWindows(vm) ? getURLSourceByDisk(vm, name) : null;
+
+export const getPVCSourceByDisk = (vm: VMKind, cdName: string) =>
```
```suggestion
export const getPVCSourceByDisk = (vm: VMKind, diskName: string) =>
```

as it is not specific for cd disks only

rawagner commented on frontend/packages/kubevirt-plugin/src/components/vm-disks/disk-summary.tsx:
```diff
@@ -0,0 +1,23 @@
+  getWindowsToolsURLByDisk,
+} from '../../selectors/vm/selectors';
+
+const getDiskSource = (vm, { name }) => {
```
this is a component so it should be 
```suggestion
const DiskSource = ({vm, disk}) => {
```
also please add types (`React.FC`...)

rawagner commented on frontend/packages/kubevirt-plugin/src/components/vm-disks/disk-summary.tsx:
```diff
@@ -0,0 +1,23 @@
+} from '../../selectors/vm/selectors';
+
+const getDiskSource = (vm, { name }) => {
+  const container = getContainerImageByDisk(vm, name);
```
I would prefer to have a function similar to https://github.com/kubevirt/web-ui-components/blob/master/src/selectors/dv/selectors.js#L17 which would replace `getContainerImageByDisk`, `getPVCSourceByDisk` and `getURLSourceByDisk`

glekner replied:
I wish I knew about this function! btw i don't see a way to determine the container image by this function, am i missing something?

rawagner replied:
there was no need. Feel free to add it :)

rawagner commented on frontend/packages/kubevirt-plugin/src/components/vm-disks/disk-summary.tsx:
```diff
@@ -0,0 +1,23 @@
+  const pvc = getPVCSourceByDisk(vm, name);
+  const url = getURLSourceByDisk(vm, name);
+
+  if (windowsToolsURL) return <li key={name}>Windows Tools - {windowsToolsURL}</li>;
```
instead of these for returns i'd use switch and one return statement.

glekner replied:
switch on what? we need to check every variable ( only one of them is not null at a time )

rawagner replied:
when you use https://github.com/kubevirt/web-ui-components/blob/master/src/selectors/dv/selectors.js#L17 as mentioned above you can do

```javascript
const diskSource = getDataVolumeSourceType(vm, name);
let summary: string;
switch(diskSource.type) {
  case DATA_VOLUME_SOURCE_URL:
    summary = `URL - ${disk.url}`;
  case DATA_VOLUME_SOURCE_PVC:
    summary = `PVC - ${diskSource.name}`;
 ...
}
return <li key={name}>{summary}</li>
```

rawagner commented on frontend/packages/kubevirt-plugin/src/components/vm-disks/disk-summary.tsx:
```diff
@@ -0,0 +1,23 @@
+  return <li key={name}>PVC - {pvc}</li>;
+};
+
+export const DiskSummary = ({ disks, vm, className }) => (
```
types

rawagner commented on frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx:
```diff
@@ -58,10 +61,17 @@ export const VMResourceSummary: React.FC<VMResourceSummaryProps> = ({ vm, canUpd
   const id = getBasicID(vm);
   const vmStatus = getVMStatus(vm, pods, migrations);
   const { launcherPod } = vmStatus;
+  const CDs = getCDRoms(vm);
```
```suggestion
  const cds = getCDRoms(vm);
```

glekner commented:
Some progress screens @matthewcarleton 
<img width="626" alt="Screen Shot 2019-09-22 at 17 43 46" src="https://user-images.githubusercontent.com/24938324/65389651-3149d800-dd61-11e9-9594-970ca509fe6e.png">
<img width="634" alt="Screen Shot 2019-09-22 at 16 44 29" src="https://user-images.githubusercontent.com/24938324/65389652-3149d800-dd61-11e9-8e87-6fed36f7382f.png">

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx:
```diff
@@ -93,6 +103,23 @@ export const VMDetailsList: React.FC<VMResourceListProps> = ({ vm, vmi, pods, mi
       </dd>
+      <dt>
+        CD-ROMs
+        {canUpdateVM && (
```
To stay consistent, the button should go to the value instead of the section description.
Please have a look at `EditButton` component within https://github.com/openshift/console/pull/2392

glekner replied:
By design, the Edit button should be by the section title. I switched to `EditButton`

mareklibra commented on frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cdrom-patches.ts:
```diff
@@ -0,0 +1,149 @@
+          },
+          spec: {
+            pvc: {
+              accessModes: [PVC_ACCESSMODE_RWM],
```
I believe `kubevirt-storage-class-defaults` ConfigMap should be used here. Including setting of the `volumeMode`. 
Please consider reuse of https://github.com/kubevirt/web-ui-components/blob/master/src/k8s/objects/pvc/pvc.js .

glekner replied:
added `volumeMode`

mareklibra replied:
Both `accessModes` and `volumeMode` should be taken from the `kubevirt-storage-class-defaults` ConfigMap

mareklibra commented on frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cdrom-patches.ts:
```diff
@@ -0,0 +1,149 @@
+              storageClassName: '',
+              resources: {
+                requests: {
+                  storage: '10Gi',
```
Why `10Gi` ?

glekner replied:
I was planning to ask about this, When I created a URL VM through the Wizard it had `10Gi` so I just copied it. Any suggestions here?

mareklibra replied:
@aglitke, any suggestion for default CDROM size?

mareklibra commented on frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cdrom-patches.ts:
```diff
@@ -0,0 +1,149 @@
+          spec: {
+            pvc: {
+              accessModes: [PVC_ACCESSMODE_RWM],
+              storageClassName: '',
```
Shouldn't the user have an option to select storage class?

glekner replied:
The design doesn't state letting the user change storage class, thoughts?

mareklibra replied:
Cc @matthewcarleton

mareklibra commented on frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cdrom-patches.ts:
```diff
@@ -0,0 +1,149 @@
+      if (type === 'url') {
+        dataVolumeTemplate = {
+          metadata: {
+            name: `${getName(vm)}-${name}`,
```
Can we use a generated name? To avoid conflicts ...

glekner replied:
`name` is a unique cd name, can't be duplicated

mareklibra commented:
@glekner , how is this related to the https://github.com/openshift/console/pull/2529 , please?

glekner commented:
@mareklibra this PR is built on top of #2529 , so its blocked by it. its the implementation of the next story in the EPIC

matthewcarleton commented:
@glekner looking great! 
are we able to implement the input group for the field/action for the `eject CD-ROM` row?
https://www.patternfly.org/v4/documentation/react/components/inputgroup/

mareklibra commented on frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cdrom-patches.ts:
```diff
@@ -0,0 +1,151 @@
+  }
+  return bootOrder;
+};
+export const getCDsPatch = (vm, cds, pvcs) => {
```
Can we add types?

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/cdrom-modal.tsx:
```diff
@@ -0,0 +1,208 @@
+      };
+      const container = getContainerImageByDisk(vm, cd.name);
+      if (container) {
+        return { ...cd, type: 'container', container };
```
please consider use of a constant from `kubevirt-plugin/src/constants/vm/storage.ts`

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/cdrom-modal.tsx:
```diff
@@ -0,0 +1,208 @@
+
+      const url = getURLSourceByDisk(vm, cd.name);
+      if (url) {
+        return { ...cd, type: 'url', url };
```
likewise

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/cdrom-modal.tsx:
```diff
@@ -0,0 +1,208 @@
+        return { ...cd, type: 'url', url };
+      }
+
+      return { ...cd, type: 'pvc', pvc: getPVCSourceByDisk(vm, cd.name) };
```
likewise

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/cdrom-modal.tsx:
```diff
@@ -0,0 +1,209 @@
+    setPVCs(getLoadedData(persistentVolumeClaims, []));
+  }, [persistentVolumeClaims]);
+
+  const onCDChange = (cdName, key, value) => {
```
types

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/vm-disks/disk-summary.tsx:
```diff
@@ -0,0 +1,66 @@
+);
+
+type DiskSummaryProps = {
+  className?: string;
```
I don't see this used

glekner replied:
could be used in the future

mareklibra replied:
so let's remove it for now. We should avoid orphaned code. Moreover, it could give false expectation when not implemented properly.

glekner replied:
removed

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/vm-disks/disk-summary.tsx:
```diff
@@ -0,0 +1,66 @@
+type DiskSummaryProps = {
+  className?: string;
+  vm: VMKind;
+  disks: any;
```
Can we use better type?

mareklibra commented on frontend/packages/kubevirt-plugin/src/selectors/vm/volume.ts:
```diff
@@ -3,3 +3,4 @@ import * as _ from 'lodash';
 export const getVolumePersistentVolumeClaimName = (volume) =>
   _.get(volume, 'persistentVolumeClaim.claimName');
 export const getVolumeDataVolumeName = (volume) => _.get(volume, 'dataVolume.name');
+export const getVolumeContainerImageName = (volume) => _.get(volume, 'containerDisk.image');
```
Per discussion in other PRs, we will replace `_.get()` use by direct access of props. Please change to:
```suggestion
export const getVolumeContainerImageName = (volume) => volume && volume.containerDisk ? volume.containerDisk.image : undefined;
```

and add types

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx:
```diff
@@ -16,24 +16,32 @@ import { VMKind, VMIKind } from '../../types';
     <>
-      <dt>{title}</dt>
+      <dt>
+        {title} <EditButton id={editButtonId} canEdit={canEdit} onClick={onEditClick} />
```
The edit button should go to value, not the title.
@matthewcarleton , @andybraren

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx:
```diff
@@ -147,6 +167,9 @@ export const VMDetailsList: React.FC<VMResourceListProps> = ({
 
 type VMDetailsItemProps = {
   title: string;
+  canEdit?: boolean;
```
These 3 props should not be needed

glekner replied:
they are if we agree to enable an optional edit button near the title and not for each row. waiting for @matthewcarleton response.

mareklibra commented on frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cdrom-patches.ts:
```diff
@@ -0,0 +1,151 @@
+      let dataVolumeTemplate: K8sResourceKind = null;
+
+      // Patches
+      if (type === 'pvc') {
```
reuse constant?

mareklibra commented on frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cdrom-patches.ts:
```diff
@@ -0,0 +1,151 @@
+      CDS = [...restOfDisks, disk];
+      VOLS = [...restOfVolumes, volume];
+      DATATEMPLATES = restOfDataTemplates;
+      if (dataVolumeTemplate) DATATEMPLATES = [...restOfDataTemplates, dataVolumeTemplate];
```
please use braces `{}`

mareklibra commented on frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cdrom-patches.ts:
```diff
@@ -0,0 +1,151 @@
+      DATATEMPLATES = restOfDataTemplates;
+      if (dataVolumeTemplate) DATATEMPLATES = [...restOfDataTemplates, dataVolumeTemplate];
+
+      if (type !== 'url')
```
please use braces `{}`

mareklibra commented on frontend/public/module/k8s/index.ts:
```diff
@@ -741,3 +741,9 @@ export type EventKind = K8sResourceKind & {
   };
 };
+
+export type CD = {
```
If i look correctly, this is used in `kubevirt-plugin` only. If so, please move it there

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/edit-button.tsx:
```diff
@@ -20,7 +20,7 @@ export const EditButton: React.FC<EditButtonProps> = (props) => {
 
 type EditButtonProps = {
-  children: any;
+  children?: any;
```
not needed, imo

glekner commented:
<img width="924" alt="Screen Shot 2019-10-03 at 12 20 59" src="https://user-images.githubusercontent.com/24938324/66114929-7929f400-e5d8-11e9-8cdf-b06e20cc6f94.png">

* Modal is now wider @matthewcarleton 
* Added Size, Storage Class inputs for URL Source CDs
* Added Size, Storage Class getters for Disks/CDs
* Implemented most of Marek's suggestion

*TODO*

* Tests

glekner commented:
- closed #2529 to centralise progress in one PR.
- removed test files, will add them to another PR.

*TODO*
Hardcode Windows Tools container to the UI
Tests

glekner commented on frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cdrom-patches.ts:
```diff
@@ -0,0 +1,169 @@
+            spec: {
+              pvc: {
+                accessModes: [PVC_ACCESSMODE_RWM],
+                volumeMode: PVC_VOLUMEMODE_FS,
```
@mareklibra where is the config map?
only place i've seen `kubevirt-storage-class-defaults` is in test files.

mareklibra replied:
It's either in `openshift-cnv` or `openshift` namespace.

Please look for `getDefaultSCAccessMode()` method which has been migrated from web-ui-components to openshift-console recently

glekner commented:
- Implemented @mareklibra suggestions but one (config map, waiting for marek's response)
- Added upstream/downstream Windows tools container names, for Windows VMs, which can now attach them as a CD.

@jelkosz I hardcoded both Windows containers, but when a user selects a container as of now, he can only select the `upstream` one, how are we planning to deal with this?
Other than that I think the PR is ready for another round of review

matthewcarleton commented:
looks great!

glekner commented:
polished css @matthewcarleton

matthewcarleton commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/_cdrom-modal.scss:
```diff
@@ -0,0 +1,32 @@
+
+.kv-cdrom-modal-restart-alert {
+  margin-bottom: 10px;
+}
```
We can remove this too. In PF we don't use fixed pixel spacing like this. I don't think we need it anyway, but in the future you could use something like `var(--pf-global--spacer--xs)`. I can chat more about Patternfly if that's helpful.

matthewcarleton commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/_cdrom-modal.scss:
```diff
@@ -0,0 +1,32 @@
+      grid-auto-flow: column;
+      grid-column-gap: var(--pf-global--gutter);
+    }
+  }
```
Could use just use the `pf-l-grid` system to do this rather than creating new classes for it? I've put this together here too: https://github.com/matthewcarleton/Openshift-Patternfly-demos/tree/master/layouts/Change-cd-modal

matthewcarleton commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/_cdrom-modal.scss:
```diff
@@ -0,0 +1,32 @@
+
+  .kv-cdrom-modal-row-name {
+    font-weight: 600;
+    margin-bottom: 5px;
```
same thing for these fixed px values as above.

matthewcarleton commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/cdrom-modal.tsx:
```diff
@@ -0,0 +1,219 @@
+              Add CD-ROM
+            </Button>
+            {cds.length > 1 && (
+              <Tooltip position="bottom" content="You have reached the maximum amount of CDs">
```
```suggestion
              <Tooltip position="bottom" content="You have reached the maximum amount of CD-ROM drives">
```

matthewcarleton commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/cdrom-modal.tsx:
```diff
@@ -0,0 +1,219 @@
+          <div className="kv-add-cd-btn">
+            <Button
+              id="vm-cd-add-btn"
+              variant="link"
```
```suggestion
              variant="link"
              isInline
```

matthewcarleton replied:
This will get the proper styles

matthewcarleton commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/cdrom-modal.tsx:
```diff
@@ -0,0 +1,219 @@
+        return { ...cd, type: StorageType.CONTAINER, container };
+      }
+
+      const url = getURLSourceByDisk(vm, cd.name);
```
When I change the type to URL it doesn't show the storage class and size, maybe an issue with my environment?

glekner commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/_cdrom-modal.scss:
```diff
@@ -0,0 +1,32 @@
+.kv-cdrom-modal-padding {
+  padding-left: 15px;
+  padding-right: 15px;
+}
```
its not related to the add-cd-rom. this padding is for each row of the modal, so it won't be close to the borders of the modal

glekner commented:
Created [2 Integration tests](https://github.com/glekner/console/blob/tests-cdrom-ui/frontend/packages/kubevirt-plugin/integration-tests/tests/vm.detail.cdrom.ts) in a different PR

1. creates new container CD, then removes it
2. creates two new container CDs, then ejects and changes them to URL, PVC

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/cdrom-modal.tsx:
```diff
@@ -0,0 +1,221 @@
+            else cd.isURLValid = true;
+          }
+        }
+        return cd;
```
Isn't this always true-ish?

glekner replied:
URL bool? yes, but if the validation boolean was false and afterwards fixed, we want to change it.

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/cdrom-modal.tsx:
```diff
@@ -0,0 +1,221 @@
+    setCDs(
+      cds.filter((cd) => {
+        if (cdName === cd.name) {
+          cd.changed = true;
```
We should not change the element within `Array.filter()`.

glekner replied:
changed array to object map, no filters anymore

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/cdrom-modal.tsx:
```diff
@@ -0,0 +1,221 @@
+
+  const onCDChange = (cdName: string, key: string, value: string) => {
+    setCDs(
+      cds.filter((cd) => {
```
What do we filter here for?

glekner replied:
changed array to object map, no filters anymore

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/cdrom-modal.tsx:
```diff
@@ -0,0 +1,221 @@
+    let index = 1;
+    let name = `cd-drive-${index}`;
+    cds.forEach((cd) => {
+      if (cd.name === name) {
```
This assumes that the `cds` array is sorted. Is it always true?

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/cdrom-modal.tsx:
```diff
@@ -0,0 +1,221 @@
+    setCDs(
+      cds.filter((cd) => {
+        if (cdName === cd.name) {
+          cd.ejected = true;
```
Let's find/filter first, then change.

glekner replied:
changed array to object map, no filters anymore

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/cdrom-modal.tsx:
```diff
@@ -0,0 +1,221 @@
+        setCDs(cds.filter((cd) => cd.name !== cdName));
+        setShowRestartAlert(true);
+      });
+    } else setCDs(cds.filter((cd) => cd.name !== cdName));
```
AFAIK, we wrap all blocks, even one-liner, into `{ }`. I see this happening on other places within this PR as well.

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/cdrom-row.tsx:
```diff
@@ -0,0 +1,202 @@
+  WINTOOLS_CONTAINER_NAMES,
+} from './constants';
+
+export const CDRomRow = ({
```
Please add types. Especially when the component is exported.

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/cdrom-row.tsx:
```diff
@@ -0,0 +1,202 @@
+          <TextInput
+            isDisabled
+            type="text"
+            id="cdrom-value"
```
There can be multiple `CDRomRow` components per page. The `id` should be unique.

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/cdrom-row.tsx:
```diff
@@ -0,0 +1,202 @@
+            value={cd[type]}
+            aria-label="cdrom-value"
+          />
+          <Button id="vm-cd-eject-btn" variant="secondary" onClick={() => onEject(name)}>
```
likewise (`id`)

glekner replied:
removed id in favour of className

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/cdrom-row.tsx:
```diff
@@ -0,0 +1,202 @@
+          <Button id="vm-cd-eject-btn" variant="secondary" onClick={() => onEject(name)}>
+            Eject CD-ROM
+          </Button>
+          <Button id="vm-cd-delete-btn" onClick={() => onDelete(name)} variant="plain">
```
unique id

glekner replied:
removed id in favour of className

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/cdrom-row.tsx:
```diff
@@ -0,0 +1,202 @@
+            <FormSelect
+              value={type}
+              onChange={(v) => onChange(name, 'type', v)}
+              id="cd-rom-modal-select-type"
```
unique `id`

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/cdrom-row.tsx:
```diff
@@ -0,0 +1,202 @@
+                id="cd-rom-modal-select-pvc"
+              >
+                {storageClasses.map((storClass) => {
+                  const className = getName(storClass);
```
nit: use `scName` instead of misleading `className`

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/vm-disks/_disk-summary.scss:
```diff
@@ -0,0 +1,10 @@
+.kubevirt-disk-summary-disk {
```
```suggestion
.kubevirt-disk-summary__disk {
```

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/vm-disks/_disk-summary.scss:
```diff
@@ -0,0 +1,10 @@
+  margin-bottom: 10px;
+
+  .kubevirt-disk-summary-disk-title {
+    font-weight: 600;
```
Isn't there a constant for that? To stay consistent ... 
Cc @matthewcarleton

matthewcarleton replied:
@glekner has this been removed? I don't see the file anymore 🤔

glekner replied:
@matthewcarleton filename changed to `disk-summary.scss`.
This file resembles the details view of the cds, not the modal

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/vms/vm-resource.tsx:
```diff
@@ -16,24 +16,32 @@ import { VMKind, VMIKind } from '../../types';
 import { getVmiIpAddressesString } from '../ip-addresses';
+import { DiskSummary } from '../vm-disks/disk-summary';
 
 import './_vm-resource.scss';
```
Please remove the underscode

mareklibra replied:
Let's not miss that comment :-)

glekner replied:
this is not my code. is it in the scope of this PR? @mareklibra

mareklibra commented on frontend/packages/kubevirt-plugin/src/k8s/patches/vm/vm-cdrom-patches.ts:
```diff
@@ -0,0 +1,171 @@
+    .filter((cd) => cd.changed)
+    .map(
+      ({ name, pvc, type, bootOrder, bus, container, windowsTools, url, storageClass, size }) => {
+        const pvcObject = pvcs.find((vol) => getName(vol) === pvc) || pvcs[0];
```
`pvcs[0]` might not be valid as the `pvcs` is passed as `persistentVolumesClaims.data` from the modal where  `persistentVolumesClaims` is optional in `CDRomModalProps`

mareklibra commented on frontend/packages/kubevirt-plugin/src/selectors/vm/selectors.ts:
```diff
@@ -96,3 +101,27 @@ export const getCloudInitUserData = (vm: VMKind) =>
+
+export const getStorageSizeByDisk = (vm: VMKind, diskName: string) =>
+  getDataVolumeStorageSize(
+    getDataVolumeTemplates(vm).find((vol) => vol.metadata.name.includes(diskName)),
```
`getName()`

glekner commented:
- Changed CDs array to a Map object (ditches filters and other array methods)
- Removed the need for the PVCs, StorageClasses in `getCDsPatch`
- Using `getDefaultSCAccessMode, getDefaultSCVolumeMode` in `getCDsPatch`
- Added Type column to the Disks tab
    - Updated Disk Row to fetch type and cdrom bus
    - Clicking on the Edit action of a CD opens the new modal
- Using `K8sResourceSelectRow` for PVCs, StorageClasses @suomiy can you review?
- Added missing Types and modal validations

@mareklibra thanks for the detailed review, can you take another look?

mareklibra commented on frontend/packages/kubevirt-plugin/src/selectors/vm/selectors.ts:
```diff
@@ -43,7 +49,7 @@ export const getWorkloadProfile = (vm: VMLikeEntityKind) =>
   findKeySuffixValue(getLabels(vmLike), TEMPLATE_FLAVOR_LABEL);
 
-export const isVMRunning = (value: VMKind) =>
+export const isVMRunning = (value: VMLikeEntityKind) =>
```
Makes no sense for `Template`, it can never be executed.

mareklibra commented on frontend/packages/kubevirt-plugin/src/selectors/vm/selectors.ts:
```diff
@@ -27,11 +33,11 @@ export const getInterfaces = (vm: VMKind, defaultValue = []): V1NetworkInterface
     : vm.spec.template.spec.domain.devices.interfaces;
 
-export const getNetworks = (vm: VMKind, defaultValue = []): V1Network[] =>
+export const getNetworks = (vm: VMLikeEntityKind, defaultValue = []): V1Network[] =>
```
This will not work for `Template`, structure differs.
Same comment for all similar changes bellow.

mareklibra commented on frontend/packages/kubevirt-plugin/src/selectors/vm/selectors.ts:
```diff
@@ -96,3 +102,27 @@ export const getCloudInitUserData = (vm: VMKind) =>
+  getVolumePersistentVolumeClaimName(getVolumes(vm).find((vol) => vol.name === diskName));
+
+export const getURLSourceByDisk = (vm: VMLikeEntityKind, name: string) =>
+  _.get(
```
All newly added selectors should use direct access (discussed in other PRs).
Example:
```
return myDVTemplate && myDVTemplate.spec && myDVTemplate.spec.source && myDVTemplate.spec.source.http;
```

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/index.tsx:
```diff
@@ -0,0 +1,38 @@
+
+type CDRomModalFirehoseProps = ModalComponentProps & {
+  vm: VMKind;
+  dataVolume?: any;
```
nit: would be better to have `DataVolumeKind` type here, but this is not a blocker.

glekner replied:
unused, removed it

mareklibra commented on frontend/packages/kubevirt-plugin/src/components/vm-disks/disk-row.tsx:
```diff
@@ -26,24 +28,28 @@ import {
 
 const menuActionEdit = (
   disk: CombinedDisk,
-  vmLikeEntity: VMLikeEntityKind,
```
It should be still possible to edit VM Template disks.
I hope I do not miss something, but is it still possible with this change?

glekner replied:
Through the disk row? i don't think its possible to see template disks there unless i'm wrong

mareklibra replied:
The same component is used for both VM- and VM Template- Disk tab. By this change and the one bellow, we will get regression for the VM Template disk edit functionality. please correct me if I am wrong.

glekner commented:
- Changed CSS @matthewcarleton (removed side paddings)
- Fixed new `ModalFooter` disabled prop and added `inProgress` prop
- Uses new ModalFooter now

glekner commented:
- Reverted changes to non-cnv files (`factory/modal`)

matthewcarleton commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/cdrom-modal.scss:
```diff
@@ -0,0 +1,24 @@
+  grid-column-gap: var(--pf-global--gutter);
+
+  .kubevirt-cdrom-modal__row-value {
+    grid-column: span 25;
```
can you explain this? Does the grid have 25 columns?

matthewcarleton commented on frontend/packages/kubevirt-plugin/src/components/modals/cdrom-vm-modal/cdrom-modal.scss:
```diff
@@ -0,0 +1,24 @@
+}
+
+.kubevirt-cdrom-modal__row {
+  display: grid;
```
I'm still not sure why we need custom classes for these. We could use `pf-l-grid` with `pf-m-gutter`. Maybe we can talk further about this if that's helpful.

glekner commented:
- Updated VM Templates to support the new CD interface
- Updated CSS Grid layout

@mareklibra  @matthewcarleton I added a button to the VM-Disks view to help create/edit CDs please approve.

![image](https://user-images.githubusercontent.com/24938324/67762134-6d054b00-fa4d-11e9-931b-048347b92ede.png)

matthewcarleton commented:
> * Updated VM Templates to support the new CD interface
> * Updated CSS Grid layout
> 
> @mareklibra @matthewcarleton I added a button to the VM-Disks view to help create/edit CDs please approve.
> 
> ![image](https://user-images.githubusercontent.com/24938324/67762134-6d054b00-fa4d-11e9-931b-048347b92ede.png)

Thanks for adding this @glekner a couple of comments:

- I'm not sure we want to have a specific action here for cd-roms. IMO it opens up a larger discussion about all the different type of disks that could exist here and having specific actions for each.  
- I think we need a separate design effort to determine how CD-ROMs are edited via the disks tab so we can consider the larger issue of editing in general here.
- We also wouldn't want to have more than one primary action here. 

Interested in @jelkosz opinion here as well.

glekner commented:
@matthewcarleton Since the table now also shows the type (which can be cdrom), thought we need a way to edit it.
I also edited the Edit button in the kebab menu for each Disk, so if its a cdrom it will open the new Modal. So I can remove the new Pimary action.

jelkosz commented:
@glekner we need to be consistent. If all the other disks are being edited using the edit from the row in which they are in, we need to open it from there. I would certainly remove the "Edit CD-ROMs button" from here.

mareklibra commented:
/lgtm
